### PR TITLE
[FEATURE] Code analysis tool

### PR DESCRIPTION
**Pull request by zethuman**
### Description
_This PR adds option to check overall quality of code._

### Issues Resolved
_Closes [#307]._ 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/code-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+on: ["push"]
+
+jobs:
+  codacy-analysis-cli:
+    name: Codacy Analysis CLI
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@main
```
Comment (dblock) [inline .github/workflows/code-analysis.yml:11]: Don't use @main, that's a moving target, pick a release.

File: CHANGELOG.md [hunk 2]
```diff
@@ -32,7 +32,8 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Adds `index_template` guide ([#289](https://github.com/opensearch-project/opensearch-go/pull/289))
 - Adds `advanced_index_actions` guide ([#288](https://github.com/opensearch-project/opensearch-go/pull/288))
 - Adds testcases to check UpdateByQuery functionality ([#304](https://github.com/opensearch-project/opensearch-go/pull/304))
-- Adds additional timeout after cluster start ([##303](https://github.com/opensearch-project/opensearch-go/pull/303))
+- Adds additional timeout after cluster start ([#303](https://github.com/opensearch-project/opensearch-go/pull/303))
+- Adds code quality analysis tool ([#313](https://github.com/opensearch-project/opensearch-go/pull/313))
```
Comment (dblock) [inline CHANGELOG.md:36]: Say "added codacy", better as a specific tool.

**Review by dblock — COMMENTED** [review]
So how does this work and what does it do? Any badges for README?

File: README.md [hunk 3]
```diff
@@ -1,4 +1,4 @@
...
...com/opensearch-project/opensearch-go/v2) [![Build](https://github.com/opensearch-project/opensearch-go/actions/workflows/lint.yml/badge.svg)](https://github.com/opensearch-project/opensearch-go/actions/workflows/lint.yml) [![Codacy Badge](https://app.codacy.com/project/badge/Grade/a6371b0f281142a0b30a9c537d9fdb7e)](https://app.codacy.com/gh/zethuman/opensearch-go/dashboard?utm_source=gh&utm_medium=referral&utm_content=&utm_campaign=Badge_grade) [![Unit](https://github.com/opensearch-project/opensearch-go/actions/workflows/test-unit.yml/badge.svg)](https://github.com/opensearch-project/opensearch-go/actions/workflows/test-unit.yml) [![Integration](https://github.com/opensearch-project/opensearch-go/actions/workflows/test-integration.yml/badge.svg)](https://github.com/opensearch-project/opensearch-go/actions/workflows/test-integration.yml) [![codecov](https://codecov.io/gh/opensearch-project/opensearch-go/branch/main/graph/badge.svg?token=MI9g3KYHVx)](https://codecov.io/gh/opensearch-project/opensearch-go) [![Chat](https://img.shields.io/badge/chat-on%20forums-blue)](https://discuss.opendistrocommunity.dev/c/clients/) ![PRs welcome!](https://img.shields.io/badge/PRs-welcome!-success)
```
[the remark this answers is not in the document]
Comment (dblock) [reply README.md:1]: 🤔 
Screenshot 2023-04-28 at 12 57 11 PM

**Comment by dblock** [comment]
I'm good with this! Do you want to try to address the above complaints from codacy before merging or leave them alone/dismiss them?

**Review by dblock — DISMISSED** [review]

**Comment by dblock** [comment]
@Jakob3xD are you ok with adding codacy here? any reasons not to?

**Comment by Jakob3xD** [comment]
> @Jakob3xD are you ok with adding codacy here? any reasons not to?

I added my concerns to the issue but in general I am okay with merging it and seeing how its works.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (zethuman) [reply README.md:1]: To be honest, I can’t understand what it wants ... 
I’m trying to figure out how to fix. But it's strange that I don't have more details, below is a screenshot of what I see. Can you share what shows in `Show more details`?

image

[the remark this answers is not in the document]
Comment (Jakob3xD) [reply README.md:1]: It show `markdownlint_MD013` which afters searching for it seems to be a line length alert. It wants a max length of 80 but has over 1000.

**Comment by zethuman** [comment]
Tasks fail due to problems on the side of the tool itself. Here is the [link](https://github.com/codacy/codacy-analysis-cli-action/issues/113) to the issue, I will follow the fix until I need to merge

**Comment by zethuman** [comment]
It seems that the problem has been solved, who has access to restart jobs, please restart

File: README.md [hunk 4]
```diff
@@ -12,7 +19,10 @@ OpenSearch Go Client
 
 ## Welcome!
 
-**opensearch-go** is [a community-driven, open source fork](https://aws.amazon.com/blogs/opensource/introducing-opensearch/) of go-elasticsearch licensed under the [Apache v2.0 License](LICENSE.txt). For more information, see [opensearch.org](https://opensearch.org/).
+**opensearch-go** is
+[a community-driven, open source fork](https://aws.amazon.com/blogs/opensource/introducing-opensearch/)
+of go-elasticsearch licensed under the [Apache v2.0 License](LICENSE.txt). For
+more information, see [opensearch.org](https://opensearch.org/).
```
Comment (dblock) [inline README.md:25]: This seems unrelated, and can we please not fold markdown?

https://code.dblock.org/2021/06/07/to-wrap-or-not-to-wrap-in-markdown.html

File: README.md [hunk 5]
```diff
@@ -31,7 +39,9 @@
 
 ## Code of Conduct
 
-This project has adopted the [Amazon Open Source Code of Conduct](CODE_OF_CONDUCT.md). For more information see the [Code of Conduct FAQ](https://aws.github.io/code-of-conduct-faq), or contact [opensource-codeofconduct@amazon.com](mailto:opensource-codeofconduct@amazon.com) with any additional questions or comments.
+This project has adopted the [Amazon Open Source Code of Conduct](CODE_OF_CONDUCT.md). For more information see the [Code of Conduct FAQ]
```
[the remark this answers is not in the document]
Comment (dblock) [reply README.md:42]: Is there a way to disable all or these?

File: README.md [hunk 6]
```diff
@@ -31,7 +39,9 @@ OpenSearch Go Client
...
```
Comment (dblock) [inline README.md:42]: This still looks wrapped.

**Comment by zethuman** [comment]
@dblock I research this topic more deeply, and as far as I understood, large tools are not particularly suitable for us. SonarQube, for example, is practically the most popular solution, but integration with SonarCloud is unlikely to please the community. And the solution with codacy actually turned out to be not so subject to customization. e.g. I could not turn off the check for markdown wrapping. So maybe we should stick with [golint](https://github.com/opensearch-project/opensearch-go/issues/307#issuecomment-1527701187) for now, as I suggested to @Jakob3xD? 
Here you can see [my experiments](https://github.com/zethuman/opensearch-go/pull/6)

The most interesting thing to note integration with github

image

**Comment by dblock** [comment]
I'm good with whatever you and @Jakob3xD can agree on!

**Comment by Jakob3xD** [comment]
@zethuman can you please rebase and maybe remove the changes from `README.md`.

**Comment by zethuman** [comment]
@Jakob3xD @dblock what about this PR? Is this acceptable or does it need to be improved?

**Comment by Jakob3xD** [comment]
The MR it self LGTM but I am concerned about the current linting issues. Will the current complains occur on every PR that gets created?

**Comment by zethuman** [comment]
@Jakob3xD @dblock can you merge this MR?

**Review by dblock — APPROVED** [review]
